### DISCUSSION TITLE
Creating CLI tools using typescript and ts-node

### DISCUSSION THREAD

**Discussion by IonelLupu** [TypeStrong/ts-node#1267] [General]
I am trying to create a CLI tool that will run typescript code instead of Javascript (I really need to run typescript and not transpile the typescript to javascript). 
I run the Typescript code by registering the `ts-node` package manually:
```js
/* index.js */
#!/usr/bin/env node
const path = require('path')

require('ts-node').register({
	transpileOnly : true,
	preferTsExts : true,
})
require('./cli')
```

```js
/* cli.ts */
const x: number = 123
console.log(x)
```

```json
/* package.json */
{
  "name": "ionellupu-test-ts-node-cli",
  "main": "index.js",
  "bin": {
    "test-ts-node-cli": "index.js"
  },
  "license": "ISC",
  "dependencies": {
    "ts-node": "^9.1.1",
    "typescript": "^4.2.2"
  }
}

```
I tested this by installing the package locally using `npm link` and it works as expected:
```sh
$ test-ts-node-cli
123
```

The problem is when I am installing the CLI tool with `npm install -g ionellupu-test-ts-node-cli` (after publishing it on npm and running `npm unlink`) and run it, I get an error: 
```bash
$ test-ts-node-cli
C:\Users\<username>\AppData\Roaming\npm\node_modules\ionellupu-test-ts-node-cli\cli.ts:1
const x: number = 123
      ^

SyntaxError: Missing initializer in const declaration
    at wrapSafe (internal/modules/cjs/loader.js:979:16)
    ...
```
This is usually shown when the code is run with the javascript engine instead of the ts-node compiler. This shows that the `require('ts-node').register()` didn't work for some reason.

I prepared a [test package](https://www.npmjs.com/package/ionellupu-test-ts-node-cli) that we can use to reproduce this problem.

One interesting this is that the script does run if I change the directory to the node_modules of the npm:
```bash
$ where test-ts-node-cli
C:\Users\<username>\AppData\Roaming\npm\test-ts-node-cli
C:\Users\<username>\AppData\Roaming\npm\test-ts-node-cli.cmd

$ cd C:/Users/<username>/AppData/Roaming/npm/node_modules

$ pwd
/c/Users/<username>/AppData/Roaming/npm/node_modules

$ test-ts-node-cli
123
```

One other interesting this is that the app works if I run it directly with `node` from `C:\Users\<username>\AppData\Roaming\npm\node_modules`:
```bash
$ pwd
/c/Users/<username>/AppData/Roaming/npm/node_modules

$ node ionellupu-test-ts-node-cli/index.js
123
```

But it doens't work if I run it from my home directory:
```bash
$ cd ~

$ pwd
/c/Users/<username>

$ node C:/Users/<username>/AppData/Roaming/npm/node_modules/ionellupu-test-ts-node-cli/index.js
C:\Users\<username>\AppData\Roaming\npm\node_modules\ionellupu-test-ts-node-cli\cli.ts:1
const x: number = 123
      ^

SyntaxError: Missing initializer in const declaration
```
It is really strange that in some cases it works and is others not.

Why is this happening and how can it be solved? 
Why isn't `ts-node.register` working properly when installing the package globally but works when using `npm link`?

**Comment by cspotcode** [comment]
At first glance, this might be caused by the default "ignore" option, which
is documented in our readme.  It could be that the file is located at a
path which is ignored, so we do not attempt to transform it.

**Reply by IonelLupu** [reply]
I am not using the `ignore` option directly using skip-ignore should fix the issue?:
```js
require('ts-node').register({
    transpileOnly : true,
    preferTsExts : true,
    skipIgnore: true,
})
```

**Reply by IonelLupu** [reply]
You, my friend, are a GOD. May the spirits be with you forever.

OMG, you just saved me 16 hours of painful debugging.

THANK YOU!!!! ❤️ 

---

So, the `ignore`  option ignores `node_modules` by default so I guess it doesn't get to require `ts-node` in the first place. Shouldn't we get an error instead? Something like `Could not require 'ts-node'` or better `Cannot call '.require' of undefined` since it doesn't load `ts-node`. 

I am thinking the app shouldn't load if the required packages don't load, right? Should I open a bug?

**Reply by cspotcode** [reply]
`ignore` means it will not transform those files.  They will still be executed by node.

In general, `node_modules` should be precompiled, which means that transformation is not required.

**Reply by IonelLupu** [reply]
I see. Interesting. Thank you once again.

I believe in a world where we will be writing Typescript and run it without transpilation to Javascript. Waiting for that moment. Long time to wait :(  Maybe someone will build a V8 for Typescript directly which should be a lot faster (due to known types). Who knows.

PS: The CLI I am building is for [Typetron](http://typetron.org/), the Node (soon Deno as well) web framework.
